hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/TestInsertDedupPolicy.scala

```diff
@@ -0,0 +1,132 @@
+    (11, "1", "rider-A", "driver-A", 1.1, 1),
+    (11, "2", "rider-B", "driver-B", 2.2, 2),
+    (10, "3", "rider-C", "driver-C", 33.90, 10),
+    (11, "5", "rider-C", "driver-C", 3.3, 3))
+
+  @ParameterizedTest
+  @MethodSource(Array("provideParams"))
+  def testInsertLogic(tableType: String,
+                      recordType: String,
+                      dedupPolicy: String): Unit = {
+    val sparkOpts: Map[String, String] = Map(
+      HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key -> "parquet",
+      HoodieWriteConfig.RECORD_MERGE_IMPL_CLASSES.key -> classOf[DefaultSparkRecordMerger].getName)
+    val fgReaderOpts: Map[String, String] = Map(
+      HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key -> "true",
+      HoodieReaderConfig.MERGE_USE_RECORD_POSITIONS.key -> "true",
+      HoodieWriteConfig.RECORD_MERGE_MODE.key -> "EVENT_TIME_ORDERING")
+    val insertDedupPolicy: Map[String, String] = Map(INSERT_DUP_POLICY.key -> dedupPolicy)
+    val opts = if (recordType.equals("SPARK")) {
+      sparkOpts ++ fgReaderOpts ++ insertDedupPolicy
+    } else {
+      fgReaderOpts ++ insertDedupPolicy
+    }
+    val columns = Seq("ts", "key", "rider", "driver", "fare", "number")
+
+    // Write the first batch of data.
+    val inserts = spark.createDataFrame(firstInsertData).toDF(columns: _*)
+    inserts.write.format("hudi").
+      option(RECORDKEY_FIELD.key, "key").
+      option(PRECOMBINE_FIELD.key, "ts").
+      option(TABLE_TYPE.key, tableType).
+      option(DataSourceWriteOptions.TABLE_NAME.key, "test_table").
+      option(HoodieCompactionConfig.INLINE_COMPACT.key, "false").
+      options(opts).
+      mode(SaveMode.Overwrite).
+      save(basePath)
+
+    val insertsWithDup = spark.createDataFrame(secondInsertData).toDF(columns: _*)
+    if (dedupPolicy.equals(FAIL_INSERT_DUP_POLICY)) {
+      // Write and check throws.
+      Assertions.assertThrows(
+        classOf[SparkException],
+        () => insertsWithDup.write.format("hudi").options(opts).mode(SaveMode.Append).save(basePath)
+      )
+    } else {
+      // Write data.
+      insertsWithDup.write.format("hudi").options(opts).mode(SaveMode.Append).save(basePath)
+      // Validate the data.
+      val df = spark.read.format("hudi").options(opts).load(basePath)
+      val finalDf = df.select("ts", "key", "rider", "driver", "fare", "number").sort("key")
+      val expected = if (dedupPolicy.equals(DROP_INSERT_DUP_POLICY)) expectedForDrop else expectedForNone
+      val expectedDf = spark.createDataFrame(expected).toDF(columns: _*).sort("key")
+      TestInsertDedupPolicy.validate(expectedDf, finalDf)
+    }
+  }
+}
+
+object TestInsertDedupPolicy {
+  def provideParams(): java.util.List[Arguments] = {
+    java.util.Arrays.asList(
+      Arguments.of("MERGE_ON_READ", "AVRO", NONE_INSERT_DUP_POLICY),
+      Arguments.of("MERGE_ON_READ", "SPARK", NONE_INSERT_DUP_POLICY),
+      Arguments.of("MERGE_ON_READ", "AVRO", DROP_INSERT_DUP_POLICY),
```

Comment:
we can avoid multiple record types. just table type and dup policy combos would do.